File: code-recipes/java/src/main/java/vendorretailprocurement/shipments/GetShipmentDetailsRecipe.java
package vendorretailprocurement.shipments;

import com.amazon.SellingPartnerAPIAA.LWAException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.threeten.bp.OffsetDateTime;
import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.vendor.shipments.v1.VendorShippingApi;
import software.amazon.spapi.models.vendor.shipments.v1.GetShipmentDetailsResponse;
import software.amazon.spapi.models.vendor.shipments.v1.Shipment;
import util.Constants;
import util.Recipe;

import java.util.List;

/**
 * Vendor Shipments API Recipe: Get Shipment Details
 * ==================================================
 * 
 * This recipe demonstrates how to retrieve shipment details using the
 * Vendor Shipments API. You can filter by date ranges, shipment status,
 * vendor shipment ID, and other criteria.
 * 
 * Use cases:
 * - Track shipment status and carrier information
 * - Get container details and tracking numbers
 * - Monitor delivery progress
 * - Retrieve shipment confirmations
 * 
 * API Operations used:
 * - getShipmentDetails: Returns shipment details based on filters
 */
public class GetShipmentDetailsRecipe extends Recipe {
    private static final Logger logger = LoggerFactory.getLogger(GetShipmentDetailsRecipe.class);

    private final VendorShippingApi vendorShippingApi = new VendorShippingApi.Builder()
            .lwaAuthorizationCredentials(lwaCredentials)
            .endpoint(Constants.BACKEND_URL)
            .build();

    @Override
    public void start() {
        // Example: Get shipments created in the last 7 days
        List<Shipment> shipments = getShipmentsByDateRange(
                OffsetDateTime.now().minusDays(7),
                OffsetDateTime.now()
        );
        logger.info("Found {} shipments", shipments.size());

        for (Shipment shipment : shipments) {
            logShipmentSummary(shipment);
        }
    }

    /**
     * Get shipment details created within a date range.
     */
    public List<Shipment> getShipmentsByDateRange(
            OffsetDateTime createdAfter,
            OffsetDateTime createdBefore) {
        try {
            GetShipmentDetailsResponse response = vendorShippingApi.getShipmentDetails(
                    50L,              // limit
                    "DESC",           // sortOrder
                    null,             // nextToken
                    createdAfter,     // createdAfter
                    createdBefore,    // createdBefore
                    null,             // shipmentConfirmedBefore
                    null,             // shipmentConfirmedAfter
                    null,             // packageLabelCreatedBefore
                    null,             // packageLabelCreatedAfter
                    null,             // shippedBefore
                    null,             // shippedAfter
                    null,             // estimatedDeliveryBefore
                    null,             // estimatedDeliveryAfter
                    null,             // shipmentDeliveryBefore
                    null,             // shipmentDeliveryAfter
                    null,             // requestedPickUpBefore
                    null,             // requestedPickUpAfter
                    null,             // scheduledPickUpBefore
                    null,             // scheduledPickUpAfter
                    null,             // currentShipmentStatus
                    null,             // vendorShipmentIdentifier
                    null,             // buyerReferenceNumber
                    null,             // buyerWarehouseCode
                    null              // sellerWarehouseCode
            );

            if (response.getPayload() != null && response.getPayload().getShipments() != null) {
                return response.getPayload().getShipments();
            }
            return List.of();
        } catch (ApiException | LWAException e) {
            logger.error("Error getting shipment details: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to get shipment details", e);
        }
    }

    /**
     * Get shipment details by vendor shipment identifier.
     */
    public Shipment getShipmentByVendorId(String vendorShipmentId) {
        try {
            GetShipmentDetailsResponse response = vendorShippingApi.getShipmentDetails(
                    1L, null, null, null, null, null, null, null, null,
                    null, null, null, null, null, null, null, null, null, null,
                    null,             // currentShipmentStatus
                    vendorShipmentId, // vendorShipmentIdentifier
                    null, null, null
            );

            if (response.getPayload() != null && response.getPayload().getShipments() != null
                    && !response.getPayload().getShipments().isEmpty()) {
                logger.info("Found shipment for vendor ID: {}", vendorShipmentId);
                return response.getPayload().getShipments().get(0);
            }
            logger.warn("No shipment found for vendor ID: {}", vendorShipmentId);
            return null;
        } catch (ApiException | LWAException e) {
            logger.error("Error getting shipment by vendor ID {}: {}", vendorShipmentId, e.getMessage(), e);
            throw new RuntimeException("Failed to get shipment by vendor ID", e);
        }
    }

    /**
     * Get shipment details by buyer reference number (Amazon's reference).
     */
    public Shipment getShipmentByBuyerReference(String buyerReferenceNumber) {
        try {
            GetShipmentDetailsResponse response = vendorShippingApi.getShipmentDetails(
                    1L, null, null, null, null, null, null, null, null,
                    null, null, null, null, null, null, null, null, null, null,
                    null, null,
                    buyerReferenceNumber, // buyerReferenceNumber
                    null, null
            );

            if (response.getPayload() != null && response.getPayload().getShipments() != null
                    && !response.getPayload().getShipments().isEmpty()) {
                logger.info("Found shipment for buyer reference: {}", buyerReferenceNumber);
                return response.getPayload().getShipments().get(0);
            }
            logger.warn("No shipment found for buyer reference: {}", buyerReferenceNumber);
            return null;
        } catch (ApiException | LWAException e) {
            logger.error("Error getting shipment by buyer reference {}: {}", buyerReferenceNumber, e.getMessage(), e);
            throw new RuntimeException("Failed to get shipment by buyer reference", e);
        }
    }

    private void logShipmentSummary(Shipment shipment) {
        logger.info("----------------------------------------");
        logger.info("Vendor Shipment ID: {}", shipment.getVendorShipmentIdentifier());
        logger.info("Buyer Reference: {}", shipment.getBuyerReferenceNumber());
        logger.info("Status: {}", shipment.getCurrentShipmentStatus());
        
        if (shipment.getTransactionType() != null) {
            logger.info("Transaction Type: {}", shipment.getTransactionType());
        }
        
        if (shipment.getSellingParty() != null) {
            logger.info("Selling Party: {}", shipment.getSellingParty().getPartyId());
        }
        
        if (shipment.getShipToParty() != null) {
            logger.info("Ship To: {}", shipment.getShipToParty().getPartyId());
        }
    }
}

File: code-recipes/python/src/recipes/vendor_retail_procurement/shipments/get_shipment_details_recipe.py
"""
Vendor Shipments API: Get Shipment Details
===========================================

Retrieves shipment details using the Vendor Shipments API.
Supports filtering by date ranges, shipment status, vendor shipment ID,
and buyer reference number.

Use cases:
- Track shipment status and carrier information
- Get container details and tracking numbers
- Monitor delivery progress
- Retrieve shipment confirmations

API operation: getShipmentDetails
"""

import logging
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Optional

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.vendor_shipments_v1.vendor_shipment_api import VendorShipmentApi

from src import config
from src.util.recipe import Recipe

logger = logging.getLogger(__name__)


class GetShipmentDetailsRecipe(Recipe):

    def __init__(
        self,
        sp_config: Optional[SPAPIConfig] = None,
        vendor_shipping_api: Optional[VendorShipmentApi] = None,
    ) -> None:
        super().__init__(config=sp_config)
        self._vendor_shipping_api = vendor_shipping_api

    @property
    def vendor_shipping_api(self) -> VendorShipmentApi:
        if self._vendor_shipping_api is None:
            client = SPAPIClient(
                self.config,
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url,
            )
            self._vendor_shipping_api = VendorShipmentApi(client.api_client)
            logger.info("VendorShipmentApi client initialized")
        return self._vendor_shipping_api

    def get_shipments_by_date_range(
        self,
        created_after: datetime,
        created_before: datetime,
        limit: int = 50,
    ) -> List[Dict[str, Any]]:
        """Get shipment details created within a date range."""
        logger.info(
            "Fetching shipments: created_after=%s, created_before=%s",
            created_after.isoformat(),
            created_before.isoformat(),
        )
        response = self.vendor_shipping_api.get_shipment_details(
            limit=limit,
            sort_order="DESC",
            created_after=created_after,
            created_before=created_before,
        )
        shipments = self._extract_shipments(response)
        logger.info("Found %d shipments", len(shipments))
        return shipments

    def get_shipment_by_vendor_id(self, vendor_shipment_id: str) -> Optional[Dict[str, Any]]:
        """Get shipment details by vendor shipment identifier."""
        logger.info("Fetching shipment by vendor ID: %s", vendor_shipment_id)
        response = self.vendor_shipping_api.get_shipment_details(
            limit=1,
            vendor_shipment_identifier=vendor_shipment_id,
        )
        shipments = self._extract_shipments(response)
        if shipments:
            logger.info("Found shipment for vendor ID: %s", vendor_shipment_id)
            return shipments[0]
        logger.warning("No shipment found for vendor ID: %s", vendor_shipment_id)
        return None

    def get_shipment_by_buyer_reference(self, buyer_reference_number: str) -> Optional[Dict[str, Any]]:
        """Get shipment details by buyer reference number."""
        logger.info("Fetching shipment by buyer reference: %s", buyer_reference_number)
        response = self.vendor_shipping_api.get_shipment_details(
            limit=1,
            buyer_reference_number=buyer_reference_number,
        )
        shipments = self._extract_shipments(response)
        if shipments:
            logger.info("Found shipment for buyer reference: %s", buyer_reference_number)
            return shipments[0]
        logger.warning("No shipment found for buyer reference: %s", buyer_reference_number)
        return None

    def start(self) -> None:
        now = datetime.now(timezone.utc)
        seven_days_ago = now - timedelta(days=7)

        shipments = self.get_shipments_by_date_range(
            created_after=seven_days_ago,
            created_before=now,
        )
        logger.info("Found %d shipments", len(shipments))

        for shipment in shipments:
            self._log_shipment_summary(shipment)

    @staticmethod
    def _extract_shipments(response: Any) -> List[Dict[str, Any]]:
        if hasattr(response, "to_dict"):
            response = response.to_dict()
        payload = response.get("payload") or {}
        return payload.get("shipments") or []

    @staticmethod
    def _log_shipment_summary(shipment: Dict[str, Any]) -> None:
        logger.info("----------------------------------------")
        logger.info("Vendor Shipment ID: %s", shipment.get("vendor_shipment_identifier"))
        logger.info("Buyer Reference: %s", shipment.get("buyer_reference_number"))
        logger.info("Status: %s", shipment.get("current_shipment_status"))
        logger.info("Transaction Type: %s", shipment.get("transaction_type"))
        selling_party = shipment.get("selling_party") or {}
        logger.info("Selling Party: %s", selling_party.get("party_id"))
        ship_to_party = shipment.get("ship_to_party") or {}
        logger.info("Ship To: %s", ship_to_party.get("party_id"))


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    recipe = GetShipmentDetailsRecipe()
    recipe.start()
